http://dev.mutt.org/trac/changeset/bd59be56c6b0

Don't mangle atime/mtime for mbox folders without new mail upon sync.
Closes #1362, #3271.

Index: mbox.c
===================================================================
--- mbox.c (revision 5916:508bfe4a2e23)
+++ mbox.c (revision 5921:bd59be56c6b0)
@@ -682,20 +682,28 @@
 /* if mailbox has at least 1 new message, sets mtime > atime of mailbox
  * so buffy check reports new mail */
-static void reset_atime (CONTEXT *ctx)
+static void reset_atime (CONTEXT *ctx, struct stat *st)
 {
   struct utimbuf utimebuf;
-  int i;
-  time_t now = time (NULL);
-
-  for (i = 0; i < ctx->msgcount; i++)
-  {
+  int i, found = 0;
+  struct stat _st;
+
+  if (!st)
+  {
+    if (stat (ctx->path, &_st) < 0)
+      return;
+    st = &_st;
+  }
+
+  utimebuf.actime = st->st_atime;
+  utimebuf.modtime = st->st_mtime;
+
+  for (i = 0; !found && i < ctx->msgcount; i++)
     if (!ctx->hdrs[i]->deleted && !ctx->hdrs[i]->read && !ctx->hdrs[i]->old)
-    {
-      utimebuf.actime = now - 1;
-      utimebuf.modtime = now;
-      utime (ctx->path, &utimebuf);
-      return;
-    }
-  }
+      found = 1;
+
+  if (found && utimebuf.actime >= utimebuf.modtime)
+    utimebuf.actime = utimebuf.modtime - 1;
+
+  utime (ctx->path, &utimebuf);
 }
 
@@ -713,4 +721,5 @@
   int first = -1;	/* first message to be written */
   LOFF_T offset;	/* location in mailbox to write changed messages */
+  struct stat statbuf;
   struct m_update_t *newOffset = NULL;
   struct m_update_t *oldOffset = NULL;
@@ -904,4 +913,13 @@
   fp = NULL;
 
+  /* Save the state of this folder. */
+  if (stat (ctx->path, &statbuf) == -1)
+  {
+    mutt_perror (ctx->path);
+    mutt_sleep (5);
+    unlink (tempfile);
+    goto bail;
+  }
+
   if ((fp = fopen (tempfile, "r")) == NULL)
   {
@@ -973,4 +991,7 @@
   }
 
+  /* Restore the previous access/modification times */
+  reset_atime (ctx, &statbuf);
+
   /* reopen the mailbox in read-only mode */
   if ((ctx->fp = fopen (ctx->path, "r")) == NULL)
@@ -998,9 +1019,4 @@
   unlink (tempfile); /* remove partial copy of the mailbox */
   mutt_unblock_signals ();
-
-  /* if mailbox has new mail, mangle atime+mtime to make buffy check
-   * report new mail for it */
-  if (!option (OPTCHECKMBOXSIZE))
-    reset_atime (ctx);
 
   return (0); /* signal success */
